include/watchdog.h

```diff
@@ -1,116 +1,168 @@
 /*
- * Copyright (c) 2015 Intel Corporation.
- *
  * SPDX-License-Identifier: Apache-2.0
  */
 
+/**
+ * @file
+ * @brief Public API for watchdog drivers.
+ */
+
+#if defined(CONFIG_WDT_LEGACY_API)
+
+/*
+ * This is the default, and will be the way until the new API below
+ * will be enforced everywhere.
+ */
+#include <watchdog_legacy.h>
+
+#else
+
 #ifndef _WDT_H_
 #define _WDT_H_
 
 #include <zephyr/types.h>
-#include <device.h>
 #include <misc/util.h>
+#include <device.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-#define WDT_MODE		(BIT(1))
-#define WDT_MODE_OFFSET         (1)
-#define WDT_TIMEOUT_MASK        (0xF)
-
-enum wdt_mode {
-	WDT_MODE_RESET = 0,
-	WDT_MODE_INTERRUPT_RESET
-};
+/**
+ * @brief Watchdog operational mode.
+ */
+#define WDT_OP_MODE_RESET		(0)
+#define WDT_OP_MODE_INTERRUPT_RESET	BIT(0)
+#define WDT_OP_MODE_MASK		(0x1)
 
 /**
- * WDT clock cycles for timeout type.
+ * @brief Watchdog behavior when CPU is in sleep state.
  */
-enum wdt_clock_timeout_cycles {
-	WDT_2_16_CYCLES,
-	WDT_2_17_CYCLES,
-	WDT_2_18_CYCLES,
-	WDT_2_19_CYCLES,
-	WDT_2_20_CYCLES,
-	WDT_2_21_CYCLES,
-	WDT_2_22_CYCLES,
-	WDT_2_23_CYCLES,
-	WDT_2_24_CYCLES,
-	WDT_2_25_CYCLES,
-	WDT_2_26_CYCLES,
-	WDT_2_27_CYCLES,
-	WDT_2_28_CYCLES,
-	WDT_2_29_CYCLES,
-	WDT_2_30_CYCLES,
-	WDT_2_31_CYCLES
-};
+#define WDT_NO_PAUSE_IN_SLEEP	(0)
+#define WDT_PAUSE_IN_SLEEP	BIT(1)
+#define WDT_PAUSE_IN_SLEEP_MASK (1 << 1)
 
+/**
+ * @brief Watchdog behavior when CPU is halted by the debugger.
+ */
+#define WDT_NO_PAUSE_HALTED_BY_DBG	(0)
+#define WDT_PAUSE_HALTED_BY_DBG		BIT(2)
+#define WDT_PAUSE_HALTED_BY_DBG_MASK	(1 << 2)
 
 /**
- * WDT configuration struct.
+ * @brief Watchdog configuration struct.
+ *
+ * timeouts - pointer to an array storing timeouts values in microseconds
+ *	      for all enabled channels. If watchdog supports multiple channels
+ *	      but only one counter reload value then the first value from the
+ *	      array will be taken.
```

Comment:
I still wonder if such high resolution for watchdog is required. Maybe it would be sufficient to use `u16_t` and milliseconds.

Reply:
My opinion is just the opposite, let use [1 us] resolution. Usually WDT is clocker from several dozen or several hundred KHZ - so resolution is always muh more than [1ms]. Regards precisin: On nRF5x it is drive form 32768 Hz and it can be the crystal oscillator, so it has precise timing. I believe some other SoC has similar capabilities - why to lock these capabilities. I see sense in keeping 1us resolution.

Reply:
Watchdog clocks are designed to be reliable not precise. Typically these are a very imprecise RC clock sources, often as a fallback when a crystal fails. But even if we had a 32768 crystal oscillator only microsecond precision is not required. We won't use watchdog to measure time. Resolution of watchdog is strictly connected with the system tick. It will never be smaller than a tick length. Watchdog timeout value lays typically between ~10ms to some seconds. It should be as small as possible but has to include all the uncertainties: inaccuracy of the source clock, scheduler. There is always safety margin added to the timeout value and it will never lay in microseconds range. If Zephyr allowed to set system tick to value smaller than 1 ms then the watchdog should do it too but it the same way kernel API does. Let's keep APIs consistent.